frontend/gateway/gateway.go

```diff
@@ -738,6 +744,450 @@ func (lbf *llbBridgeForwarder) Inputs(ctx context.Context, in *pb.InputsRequest)
+			eg.Go(func() error {
+				<-initCtx.Done()
+				pioMu.Lock()
+				pio := pios[pid]
+				delete(pios, pid)
+				pioMu.Unlock()
+				if pio != nil {
+					err = pio.Close()
+				}
+				return nil
+			})
+
+			process := &executor.ProcessInfo{
+				Meta: executor.Meta{
+					Args: init.Meta.Args,
+					Env:  init.Meta.Env,
+					User: init.Meta.User,
+					Cwd:  init.Meta.Cwd,
+					Tty:  init.Tty,
+				},
+				Stdin:  pio.fds[0],
+				Stdout: pio.fds[1],
+				Stderr: pio.fds[2],
+				Resize: pio.resize,
+			}
+			process.Meta.Env = addDefaultEnvvar(process.Meta.Env, "PATH", utilsystem.DefaultPathEnv)
+			if init.Tty {
+				process.Meta.Env = addDefaultEnvvar(process.Meta.Env, "TERM", "xterm")
+			}
+
+			// save process io for fd/resize events later
+			pioMu.Lock()
+			// ensure we are not clobbering something else in parallel
+			if _, ok = pios[pid]; ok {
+				pioMu.Unlock()
+				return status.Errorf(codes.AlreadyExists, "Process %s already exists", pid)
+			}
+			pios[pid] = pio
+			pioMu.Unlock()
+
+			for fd, file := range pio.fds {
+				file := file
+				// don't read from container stdin
+				if fd == 0 || file == nil {
+					continue
+				}
+				eg.Go(func() error {
+					dest := &outputWriter{
+						stream:      srv,
+						fd:          uint32(fd),
+						processID:   pid,
+						containerID: id,
+					}
+					_, err := io.Copy(dest, file)
+					if err == nil {
+						// no error so must be EOF
+						err = srv.Send(&pb.ExecMessage{
+							ProcessID: pid,
+							Input: &pb.ExecMessage_File{
+								File: &pb.FdMessage{
+									Fd:  uint32(fd),
+									Eof: true,
+								},
+							},
+						})
+					}
+					return err
+				})
+			}
+
+			// Send the Started message with the pid.  We send it now so that it will
+			// arrive to the client before any Fd messages that might occur async when
+			// the container starts up.
+			err = srv.Send(&pb.ExecMessage{
+				ProcessID: pid,
+				Input: &pb.ExecMessage_Started{
+					Started: &pb.StartedMessage{},
+				},
+			})
+			if err != nil {
+				return err
+			}
+
+			eg.Go(func() error {
+				defer initCancel()
+
+				err := lbf.execProcess(ctx, ctr, id, process)
+				var status uint32
+				var exitError *executor.ExitError
+				if errors.As(err, &exitError) {
+					status = exitError.ExitCode
+				}
+				logrus.Debugf("Process %s exited with status %d, error=%s", pid, status, err)
+
+				sendErr := srv.Send(&pb.ExecMessage{
+					ProcessID: pid,
+					Input: &pb.ExecMessage_Exit{
+						Exit: &pb.ExitMessage{
+							Code: status,
+						},
+					},
+				})
+
+				if sendErr != nil && err != nil {
+					return errors.Wrap(sendErr, err.Error())
+				} else if sendErr != nil {
+					return sendErr
+				}
+
+				if err != nil && status != 0 {
+					// this was a container exit error which is "normal" so
+					// don't return this error from the errgroup
+					return nil
+				}
+				return err
+			})
+		}
+	}
+}
+
+func addDefaultEnvvar(env []string, k, v string) []string {
```

Comment:
this was copied from ops/exec.go, we can move it to a common util package if desired, was not sure where to put it.